llama_stack/apis/dataset/dataset.py

```diff
@@ -4,57 +4,124 @@
 # This source code is licensed under the terms described in the LICENSE file in
 # the root directory of this source tree.
 
+from abc import ABC, abstractmethod
 from enum import Enum
-from typing import Any, Dict, Optional, Protocol
-
-from llama_models.llama3.api.datatypes import URL
+from typing import Any, Dict, Generic, Iterator, Literal, Protocol, TypeVar, Union
 
 from llama_models.schema_utils import json_schema_type, webmethod
+from llama_models.llama3.api.datatypes import *  # noqa: F403
+
+from pydantic import BaseModel, Field
+from typing_extensions import Annotated
+
+# A sample (row) from dataset
+TDatasetSample = TypeVar("TDatasetSample")
+
+
+@json_schema_type
+class DatasetSample(BaseModel): ...
 
-from pydantic import BaseModel
+
+@json_schema_type
+class DictSample(DatasetSample):
+    data: Dict[str, Any]
+
+
+# A sample (row) from evals intermediate dataset
+TProcessedSample = TypeVar("TProcessedSample")
 
 
 @json_schema_type
-class TrainEvalDatasetColumnType(Enum):
-    dialog = "dialog"
-    text = "text"
-    media = "media"
-    number = "number"
-    json = "json"
+class PredictionSample(BaseModel):
+    completion_message: str
 
 
 @json_schema_type
-class TrainEvalDataset(BaseModel):
-    """Dataset to be used for training or evaluating language models."""
+class ProcessedDictSample(DictSample):
```

Comment:
class PreprocessedOutputDictSample()
    model_input: 
    expected_output: (for scoring)

class PredictionOutputDictSample()
    prediction_output: 
    expected_output: (for scoring)

class PostProcessedOutput()